docs/src/specification.md

```diff
@@ -1,32 +1,279 @@
-esse, obstat gravitate. Obscura atque coniuge, per de coniunx, sibi **medias
-commentaque virgine** anima tamen comitemque petis, sed. In Amphion vestros
-hamos ire arceor mandere spicula, in licet aliquando.
+YARTBML is a functional programming langauge with a set of basic primitive and complex datatypes. It is a statically typed general purpose language that runs with its own interpreter built in GO. Memory management is handled by GO, and the interpreter is a binary file which was compiled by Go to the specific machine architecture.
 
-```java
-public class Example implements LoremIpsum {
-	public static void main(String[] args) {
-		if(args.length < 2) {
-			System.out.println("Lorem ipsum dolor sit amet");
-		}
-	} // Obscura atque coniuge, per de coniunx
-}
+# 2 Lexical Structure
+This section specifies the lexical structure of the programming language.
+
+### 2.1 Keywords
+The language includes reserved keywords that have special meanings and cannot be used as identifiers.
+
+```
+<keyword> ::= 
+          "let" 
+	| "fn" 
+	| "return" 
+	| "if" 
+	| "else" 
+	| "true" 
+	| "false"
+```
+
+### 2.2 Integer Literals
+Integers are sequences of digits
+```
+<integer> ::= <digit>+
+```
+
+\pagebreak
+
+### 2.3 String Literals
+Strings are sequences of characters enclosed in double quotes.
+```
+<string> ::= "\"" <char>* "\""
+<char> ::= 
+	<any character except newline or double quote>
+```
+
+### 2.4 Delimiters
+Delimiters separate tokens in the code.
+```
+<delimiter> ::= 
+	  "(" 
+	| ")" 
+	| "{" 
+	| "}" 
+	| "[" 
+	| "]" 
+	| "," 
+	| ";" 
+	| ":"
+```
+
+### 2.5 Operators
+Operators are symbols used to perform operations on values.
+```
+<operator> ::= 
+	  "+" 
+	| "-" 
+	| "*" 
+	| "/" 
+	| "==" 
+	| "!=" 
+	| "<" 
+	| ">" 
+	| "<=" 
+	| ">="
+ ```
+
+\pagebreak
+
+### 2.6 Identifiers
+Identifiers are sequences of letters, digits, and underscores that do not start with a digit.
+```
+<identifier> ::= 
+	<letter> (<letter> 
+				| <digit> 
+				| "_")*
+<letter> ::= 
+	  "a" 
+	| "b" 
+	| ... 
+	| "z" 
+	| "A" 
+	| "B" 
+	| ... 
+	| "Z"
+<digit> ::= 
+	  "0" 
+	| "1" 
+	| ... 
+	| "9"
+```
+
+### 2.7 White Space
+Whitespace characters include spaces, tabs, and newline characters and are used to separate tokens and improve code readability.
+```
+<whitespace> ::= 
+	  <space> 
+	| <tab> 
+	| <newline>
+<space> ::= 
+	" "
+<tab> ::= 
+	"\t"
+<newline> ::= 
+	"\n"
+```
+
+\pagebreak
+
+# 3 Grammar
+This section specifies the grammar of the language
+
+### 3.1 Binding Values
+The YARTBML REPL allows users to bind values to names using the let statement.
+```
+<let_statement> ::= 
+	"let" <identifier> "=" <expression> ";"
+```
+
+### 3.2 Supported Data Types
+In addition to integers, booleans, and strings, YARTBML supports arrays and hashmaps.
+
+### 3.3 Binding Arrays
+Arrays of integers can be bound to names using the following syntax:
+```
+<let_statement> ::= 
+	"let" <identifier> "=" "[" <array_elements> "]" ";"
+<array_elements> ::= 
```

Comment:
There should be an array_literal ebnf definition to declare the array datatype as its own, akin to integer literal.

```
<array_literal> ::= "[" [ <element_list> ] "]"
<element_list> ::= <element> ( "," <element> )*
<element> ::= <expression>
```